src/cudamatrix/cu-allocator.cc

```diff
@@ -123,14 +127,68 @@ void* CuMemoryAllocator::MallocPitchInternal(size_t row_bytes,
   return ans;
 }
 
+
+std::string GetFreeGpuMemory(int64* free, int64* total) {
+#ifdef _MSC_VER
+  size_t mem_free, mem_total;
+  cuMemGetInfo_v2(&mem_free, &mem_total);
+#else
+#if (CUDA_VERSION >= 3020)
+  // define the function signature type
+  size_t mem_free, mem_total;
+#else
+  unsigned int mem_free, mem_total;
+#endif
+  {
+    // we will load cuMemGetInfo_v2 dynamically from libcuda.so
+    // pre-fill ``safe'' values that will not cause problems
+    mem_free = 1; mem_total = 1;
+    // open libcuda.so
+    void* libcuda = dlopen("libcuda.so", RTLD_LAZY);
+    if (NULL == libcuda) {
+      KALDI_WARN << "cannot open libcuda.so";
+    } else {
+      // define the function signature type
+      // and get the symbol
+#if (CUDA_VERSION >= 3020)
+      typedef CUresult (*cu_fun_ptr)(size_t*, size_t*);
+      cu_fun_ptr dl_cuMemGetInfo = (cu_fun_ptr)dlsym(libcuda,"cuMemGetInfo_v2");
+#else
+      typedef CUresult (*cu_fun_ptr)(int*, int*);
+      cu_fun_ptr dl_cuMemGetInfo = (cu_fun_ptr)dlsym(libcuda,"cuMemGetInfo");
+#endif
+      if (NULL == dl_cuMemGetInfo) {
```

Comment:
This is subtle and doesn't really matter, but the *right* way to do this is to call dlerror() and check if the return value is NULL. This is because it is possible that a shared object may have a symbol which is initialized to NULL. Of course, that never happens in this case.